### PR TITLE
Initial work

### PR DESCRIPTION
**Pull request by sonnysideup**
This is a library that can be re-used in Ranchhand, Shipyard, and any future projects that need to orchestrate Helm operations.

The helm source code currently is not very user-friendly when trying to use it as a client library; perhaps this will change in the future. Until then, we can build out a _simple client wrapper_ here and include optional support for AppR and ChartMuseum.

### CODE REVIEW

File: extra.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
+package helmexec
+
+import "strings"
+
+func (w Wrapper) IsRelease(name string) bool {
+	out, err := w.exec("list", "--all", "--short", name)
```
Comment (steved) [inline extra.go]: This type of stuff could presumably use `--output yaml`?

File: plugin.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+	if err != nil {
+		return
+	}
+
+	re := regexp.MustCompile(`(?m)^(\S+)\s+(\S+)\s+(.*)$`)
+	matches := re.FindAllStringSubmatch(string(out), -1)
+	header := []string{"NAME", "VERSION", "DESCRIPTION"}
+	if matches == nil || !reflect.DeepEqual(strings.Fields(matches[0][0]), header) {
+		return nil, fmt.Errorf("invalid list format: %s", string(out))
+	}
```
Comment (steved) [inline plugin.go]: oof

File: repo_test.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+					Username: "missing-pass",
+				},
+				[]string{"helm", "repo", "add", repoName, repoURL},
+			},
+			{
+				&he.RepoAddOptions{
+					Password: "missing-user",
+				},
+				[]string{"helm", "repo", "add", repoName, repoURL},
+			},
```
Comment (steved) [inline repo_test.go]: pretty slick

File: wrapper.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+		runner:     command.ShellRunner{},
+	}
+}
+
+func (w *Wrapper) SetRunner(runner command.Runner) error {
+	if runner != nil {
+		w.runner = runner
+		return nil
+	}
+	return errors.New("command runner cannot be nil")
```
Comment (steved) [inline wrapper.go]: Is this pattern common? Seems almost overly defensive for golang

File: repo.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+	}
+
+	re := regexp.MustCompile(`(?m)^(\S+)\s+(.*)$`)
+	matches := re.FindAllStringSubmatch(string(out), -1)
+	header := []string{"NAME", "URL"}
+	if matches == nil || !reflect.DeepEqual(strings.Fields(matches[0][0]), header) {
+		return nil, fmt.Errorf("invalid list format: %s", string(out))
+	}
+
+	for _, data := range matches {
```
Comment (steved) [inline repo.go]: why not `matches[1:]`?

File: extra.go (hunk 1 above)
Comment (sonnysideup) [reply extra.go]: Totally. I wish they had that option. Perhaps we should open an issue seeing if they would accept a PR for that functionality?

File: plugin.go (hunk 2 above)
Comment (sonnysideup) [reply plugin.go]: I know, this really hurts. Any ideas about how to better parse this non-machine output?

File: repo.go (hunk 5 above)
Comment (sonnysideup) [reply repo.go]: Good catch, I missed that bit...derp! I think a lot of the guts of both `RepoList` and `PluginList` can be refactored into an unexported function.

File: repo_test.go (hunk 3 above)
Comment (sonnysideup) [reply repo_test.go]: 😎

File: wrapper.go (hunk 4 above)
Comment (sonnysideup) [reply wrapper.go]: I thought about exporting the `runner` field but that allows `nil` values. That's a simpler implementation but it _did_ bite me in the ass during testing. I'm on the fence really.

File: repo.go (hunk 5 above)
Comment (sonnysideup) [reply repo.go]: remedied with a589941

File: extra.go (hunk 1 above)
Comment (steved) [reply extra.go]: `helm list` seems to be the only one that _actually_ does have this option oddly enough

File: plugin.go (hunk 2 above)
Comment (steved) [reply plugin.go]: Nope... and welp: https://github.com/helm/helm/issues/4266#issuecomment-410407630

File: wrapper.go (hunk 4 above)
Comment (steved) [reply wrapper.go]: No problem, nothing wrong with it I'm just curious

File: plugin.go (hunk 2 above)
Comment (sonnysideup) [reply plugin.go]: So basically I can use this in `IsRelease()` and nothing else right. Le sigh. 🙄

**Pull request closed without merging**